### ISSUE TITLE
ENH/DEP: Clean up quad return types

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#18574]
The returns for several quadrature methods are from a time long-gone by, when switching output signature based on arguments was considered acceptable. Also, there are other issues to fix, like not returning enough information in all cases.

Quoting @WarrenWeckesser from #7816:
> The return values of `quad` is another one of those ugly warts in the scipy API.
> 
> The number of return values depends on the argument `full_output`, and on whether or not the algorithm converges.

As of 1.10.1, the returns looks like this (though it's not the fully story even):
https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/integrate/_quadpack_py.py#L505-L513

#7816 was tackled in #18298 in an output-compatible way, i.e. returns would have retained their shape as a 5-tuple if `full_output=True` and a 2-tuple if `full_output=False` (even though enhanced by being a `Bunch` that just fakes being a 2-tuple), where the `full_output=True` case would have looked something like this:
```
Real:    QuadResult(integral, abserr, infodict=infodict, message=message, explain=explain)
Complex: QuadResult(integral, abserr, infodict={"real": real_infodict, "imag":imag_infodict}, message={"real": real_message, "imag":imag_message}, explain={"real": real_infodict, "imag":imag_infodict})
```

I think this is disadvantageous on several points, e.g. the nesting of the dicts for complex, the divergence between real & complex, the massive extent of infodict. #18298 ended up not being merged for 1.11, and some of the other work that had happened already (#17837, #17964) was reverted last-minute in #18560. I'm grateful for @mdhaber and @j-bowhay for allowing more time on this topic, and apologize for delaying this. 

I believe we should use our "breakage budget" to change this API _once_ for the foreseeable future, and not introduce something now that we're not happy with for the next 5-10 years at least.

My thoughts on the API design are as follows:
* We need to move away completely from mutable return signature, even if we need to wait until SciPy 2.0 to remove it completely.
  * This can be done by introducing a `legacy` keyword that starts as being `True` (but raising a deprecation warning) and does
    ```
    if legacy:
        return ...  #  as current signature: y, abserr, infodict, message, explain
    else:
        return ...  # new signature, to be determined
    ```
  * we can change the default to `legacy=False` after 2 releases
* shape of the result type: dictionary wrapped in a special class? This is the `Bunch`-approach, but perhaps can be changed further
  * One approach is to package _everything_ into (say) `QuadResult`, including the integral itself. This is what #18298 was doing, and is a fine approach
  * My suspicion however is that >90% of the use-cases for `quad` are still just "get me the integral", so I think it might be interesting to consider a signature:
    ```
    return y, QuadInfo(...)
    ```
    which would allow people to easily discard the information if they don't need it (e.g. `result, _ = quad(...)`), rather than having to learn how to extract this from `QuadInfo` (if they're not interested in it).
* Dealing with the complex case. I think we should have _the same_ structure in `QuadResult` /  `QuadInfo` regardless of whether we have real or complex integrals. The choice then comes down to whether we nest the imaginary part as an optional field in the result, or whether we do something like `.real` / `.imag`, where real is always populated, and `.imag` might be missing.
  * Option A:
    ```
    QuadResult(..., imag=QuadResult, ...)
    ```
  * Option B:
    ```
    QuadResult(.real.<key>=value, .imag.<key>=value)
    ```
  * Option C: something else?
* fully unpack & enrich content - in particular, I think we should not be giving back error flags anywhere.
  * For example, we should be able to replace the following with something better:
    > `ierlst`: A rank-1 integer array of length ``M_f`` containing an error flag corresponding to the interval in the same position in `infodict['rslist']`.  See the explanation dictionary (last entry in the output tuple) for the meaning of the codes. 
  * I'd also like to get rid of the following kind of instructions. I think we should just compute useful information and make it easily accessible. If we want to save performance and write a separate result class, we can even do this lazily.
    > Let I be the sequence `infodict['iord']` and let E be the sequence `infodict['elist']`.  Then `E[I[1]], ..., E[I[L]]` forms a decreasing sequence.
* size/structure of current `infodict`. The current infodict is a mess of different fields which are all pretty illegible. I would fundamentally reshuffle these and provide a nicer interface for them. @mdhaber had the idea of not packing everything from infodict into our new result type right away, but staying minimal and seeing if people complain that something's missing. That sounds like a good approach to me. Besides the question of when or if we introduce given fields in the result, I'd like to organize them in ways that make sense based on their information (e.g. about the nodes, the weights, the intervals). 

I came up with a quick draft in #18298 from looking at the docstrings (adapted slightly here), but I'm certain this can and should be refined. It's just here for illustrative purposes for now, all the sections and their content are up for bikeshedding.
```
QuadInfo: {
    # errors (should this be a dict or part of `'integration'`?)
    'abserr': ...
    # basics about eval
    'integration': {
      'neval': ...,
      'last': ...,
      # perhaps enriched by...
      'routine': # name of the quadpack routine that was called
      <crucial input kwargs>
    },
    # nested QuadResult for imaginary component, if complex_func
    'imag': None or QuadResult,
    # details
    'subintervals': {
        # all relevant information about subintervals (suitably renamed if we want),
        # e.g. replacing alist, blist, rlist, elist, iord for finite integration
        # resp. lst, rslst, erlst, ierlst for infinite integration
    },
    'break_points': None or {
        # equivalents of pts, level, ndin, if points kwarg was provided (else None)
    },
    'weights': None or {
        # reflect weight & wvar kwargs for completeness (if provided, else the whole thing is None)
        # plus extra info for sin/cos weights, if applicable.
    },
    'moments': None or {
        # momcom, nnlog, chebmo
    }
}
```

**Comment by ev-br** [comment]
cross-ref https://github.com/scipy/scipy/issues/18566#issuecomment-1567022617 for an alternative strategy w.r.t. quadpack. Note that I'm certainly not trying to block the effort suggested here! All I'm saying is that I believe there is an alternative with potentially better ROI.

**Comment by j-bowhay** [comment]
> #7816 was tackled in #18298 in an output-compatible way, i.e. returns would have retained their shape as a 5-tuple if `full_output=True` and a 2-tuple if `full_output=False` (even though enhanced by being a `Bunch` that just fakes being a 2-tuple), where the `full_output=True` case would have looked something like this:
> 
> ```
> Real:    QuadResult(integral, abserr, infodict=infodict, message=message, explain=explain)
> Complex: QuadResult(integral, abserr, infodict={"real": real_infodict, "imag":imag_infodict}, message={"real": real_message, "imag":imag_message}, explain={"real": real_infodict, "imag":imag_infodict})
> ```

To clarify the behavior of #18298 is:
- `full_output=True` and  `full_output=False`: return something that is BC compatible and raise a deprecation warning.
- `full_output=None`  (default value): return results object eg. what we are discussing here.

With this we can remove mutating outputs in two releases.

> I think this is disadvantageous on several points, e.g. the nesting of the dicts for complex, the divergence between real & complex, the massive extent of infodict. 

Agree the nested dicts and large infodict are not necessarily the nicest. The divergence between real & complex seems somewhat unavoidable given they are calculated in a different manner.

> * We need to move away completely from mutable return signature, even if we need to wait until SciPy 2.0 to remove it completely.
>   
>   * This can be done by introducing a `legacy` keyword that starts as being `True` (but raising a deprecation warning) and does
>     ```
>     if legacy:
>         return ...  #  as current signature: y, abserr, infodict, message, explain
>     else:
>         return ...  # new signature, to be determined
>     ```
>   * we can change the default to `legacy=False` after 2 releases
> * shape of the result type: dictionary wrapped in a special class? This is the `Bunch`-approach, but perhaps can be changed further
>   
>   * One approach is to package _everything_ into (say) `QuadResult`, including the integral itself. This is what [DEP: integrate.quad: deprecate argument `full_output` #18298](https://github.com/scipy/scipy/pull/18298) was doing, and is a fine approach

I personally don't think we need to go down the legacy keyword. At least in my mind, the only sticking point to #18298 was the structure of the result object.

>   * My suspicion however is that >90% of the use-cases for `quad` are still just "get me the integral", so I think it might be interesting to consider a signature:
>     ```
>     return y, QuadInfo(...)
>     ```    
>     which would allow people to easily discard the information if they don't need it (e.g. `result, _ = quad(...)`), rather than having to learn how to extract this from `QuadInfo` (if they're not interested in it).

I am -1 on this. The result object pattern is pretty ubiquitous in SciPy.

> * Dealing with the complex case. I think we should have _the same_ structure in `QuadResult` /  `QuadInfo` regardless of whether we have real or complex integrals. The choice then comes down to whether we nest the imaginary part as an optional field in the result, or whether we do something like `.real` / `.imag`, where real is always populated, and `.imag` might be missing.

I think we should aim for consistency across `quad`,  `nquad` and `quad_vec`. Having `.real` / `.imag` seems a little cumbersome to me as 99% of people are probably working with real integrals.

**Comment by h-vetinari** [comment]
> cross-ref [#18566 (comment)](https://github.com/scipy/scipy/issues/18566#issuecomment-1567022617) for an alternative strategy w.r.t. quadpack.

From there:
> One other example is QUADPACK. Maybe the best course of action is to assess what's missing in `quad_vec`, port the missing bits; once it's feature complete, we can just deprecate `quad`.

If people are comfortable with that, sure. I think `integrad.quad` is a nicer name than `integrate.quad_vec`, but perhaps we can reintroduce it as a wrapper later.

> Note that I'm certainly not trying to block the effort suggested here! All I'm saying is that I believe there is an alternative with potentially better ROI.

Yeah, this is still mostly explorative.

**Comment by h-vetinari** [comment]
> >   ```
> >   return y, QuadInfo(...)
> >   ```
> 
> I am -1 on this. The result object pattern is pretty ubiquitous in SciPy.

I don't feel very strongly about this, just though I'd bring it up. We should then however minimize the default `__repr__` of `QuadResult` so as to not blow up the default return in a REPL from just doing `quad(...)`.

> > Dealing with the complex case. 
> 
> I think we should aim for consistency across `quad`, `nquad` and `quad_vec`. Having `.real` / `.imag` seems a little cumbersome to me as 99% of people are probably working with real integrals.

Agreed about consistency and about most integrals being real. That's why I suggested the complex output to be a field in `QuadResult`, which I think is better than `.real` / `.imag` everywhere.

**Comment by h-vetinari** [comment]
@mdhaber 
Could you share your thoughts on this perhaps?

**Comment by mdhaber** [comment]
I started working on the return signature of `quad` before two important developments.

1) After working on gh-18650, I'm thinking that `quadrature`, `fixed_quad`, and `romberg` can be marked as legacy. Even for the easiest integrals, `_tanhsinh` speed is on par with or faster than these routines, and it's much more reliable and fully-featured (infinite integration limits, handles endpoint singularities, modern input validation, etc.). There may be room for these quadrature *rules* in SciPy, but once gh-18650 is done and vectorized, adding these rules into the framework of `_tanhsinh` would likely be very easy. Then we would have versions of these rules with all the advantages noted at the top of gh-18650 (log-integration, vectorization, GPU speedup), all with a shared infrastructure (to reduce maintenance burden) and consistent interface.

2) gh-18566 proposes replacing `quad` with a translation in a more maintainable language. (I'm not sure that `quad_vec` is a good start along these lines - it's slower and less reliable than `quad`, and it appears to eclipsed even by `_tanhsinh`. So I can see it being declared legacy, too, perhaps after some interval splitting logic is added on top of `_tanhsinh`.). `_tanhsinh` (which was drafted in a day primarily to be an integrator for private use within `stats`) cannot completely eclipse QUADPACK, so I think there is room for this. If this is done, then there is potential to permit complex dtypes, probably with minimal algorithmic adjustments. This would eliminate the need to split the real and complex parts of the result object.

So maybe my thoughts now are to see how gh-18650 and gh-18566 play out before working on a backward compatible revision to the output of `quad`. In the end, maybe we'd have one general-purpose integrator that is a derivative of QUADPACK, and a homebrew integrator descended from `_tanhsinh`. The former would be best for individual integrals, especially those that are tricky and have costly function evaluations; the latter would be good in most cases but also have all the other features of `_tanhsinh`. It is always easier to start with a clean slate than to go through deprecations, and when this much going on, it could easily be justified.

**Comment by h-vetinari** [comment]
I don't mind waiting a bit, though we do have an open issue about completeness of the return type that we postponed from 1.11.

What I'd like to do (without having a specific plan), is to at least deprecate the variable return type already (unless we expect to fully remove all those methods). That part will need to change in any case, but of course, that deprecation generally needs to provide an alternative.

So I'm ambivalent about waiting, but certainly not opposed.

**Comment by mdhaber** [comment]
If you can translate Fortran, then we don't need to wait. Even if not, we can support the other Fortran translations to free the time of those who can. I guess we have already done done of that in gh-18225 and gh-18358. I'll continue by reviewing gh-18679.

**Comment by h-vetinari** [comment]
Just thought of this today, I think we should try to tackle this in the 1.12 timeframe.

**Comment by mdhaber** [comment]
Given the merge of gh-18650 and progress toward gh-18566, one suggestion would be to deprecate `quadrature` in 1.12, and use the function name when it's available (1.14) to provide a unified public interface for `tanhsinh` and a translated QUADPACK.

**Comment by h-vetinari** [comment]
If we do that, what would be the transition? We usually try to provide a clear replacement, and I think repurposing the name directly after it becomes free again is also fraught with peril, because it'll catch people who didn't respond to the warning with unexpected errors (due to wrong API usage) rather than a clear "this function isn't here anymore".

I think it might be more feasible (even though it's more work) to provide a glue layer that keeps old and new working under `quadrature` (e.g. with some form of `legacy=True` switch), and then eventually remove it.

**Comment by h-vetinari** [comment]
> I'd think that in 1.12, users of `quadrature` would get a deprecation message that advises the switch to `quad`.

That sounds fine to me then, along with the rest of the plan. 👍

**Comment by mdhaber** [comment]
Here is some support for deprecating `quadrature` and `romberg` with `quad` as the replacement.

I benchmarked the three integrators on 818 problems from "boost-test.xlsx", which I found in a discussion of Boost math's tanh-sinh here: https://github.com/boostorg/math/issues/706#issuecomment-943934101. With the default `rtol=1.49e-8`, here is a histogram of the relative errors. (Note that 670 of the integrals have a true magnitude between 0.1 and 10, so whether we consider absolute or relative error doesn't matter much.)

![image](https://github.com/scipy/scipy/assets/6570539/d8373509-bcb5-4643-b9c8-b276179d4dfe)

---

`romberg` produces NaN in 266 cases, and it satisfies the requested tolerance in less than half (405). It doesn't provide an error estimate, so there's no indication that the result is incorrect except warnings or NaN output. 

This plot shows the wall clock time for the problems in which `romberg` satisfies the the requested tolerance.
![image](https://github.com/scipy/scipy/assets/6570539/315e5cd6-b74f-44e9-b7c2-78abca4091a8)

So it's quite fast when it works. But it tends to require more function evaluations than `quad`.

![image](https://github.com/scipy/scipy/assets/6570539/21e1d4b3-7e72-49cf-8505-e6d5e544f9a5)

So that speed advantage is just because I have `vec_func` set to True. With the default `vec=False`:

![image](https://github.com/scipy/scipy/assets/6570539/df95efcc-089c-4cfe-9303-b6eaed17cccd)

Speed aside, given that there's a 50/50 chance of producing garbage output, I don't think that's something we want to be public.

---

`quadrature` only produces 2 NaNs, but it still only satisfies the requested tolerance for about half the problems (425), and its error estimate is greater than the true error in 307 of those cases. 

Even with `vec_func=True`, it's not faster than `quad`.

![image](https://github.com/scipy/scipy/assets/6570539/a3e11b32-2c8f-44e9-a2c5-e969e9bc32f7)

![image](https://github.com/scipy/scipy/assets/6570539/b026ed30-eead-46db-8c94-1c30f6fc222b)

and of course it's much slower than `quad` with the default `vec_func=False`.

![image](https://github.com/scipy/scipy/assets/6570539/8a5e01c8-bb5c-4013-bb06-7afa5374fd51)

---

`quad` achieves the requested `rtol` in almost all (794) cases, and only occasionally underestimates its own error (26 cases total). For the cases `quadrature` and `romberg` integrate to the required accuracy, `quad` tends to require fewer function evaluations, and unless the user opts into the advangtage of vectorized function evaluations, `quad` tends to be faster.

So I'd propose that we deprecate `quadrature` and `romberg`, directing users to `quad` instead. When the `quadrature` name is free, we can use it to provide a unified interface for `_tanhsinh` and a Python version of `quad`.

**Comment by h-vetinari** [comment]
Great analysis! `quad` looks better in basically all aspects, so the plan sounds good to me!

**Comment by mdhaber** [comment]
Sounds good. Are you going to champion this? The deprecation would need an email to the mailing list. If it is not disapproved, I can review/merge a deprecation PR.

**Comment by h-vetinari** [comment]
I can write the mail to the mailing list (I have to write another deprecation related one anyway), but I'm not sure I'm going to be able to champion this. I can perhaps make the PR for the initial deprecation though (but generally quite underwater with other commitments ATM)

**Comment by mdhaber** [comment]
OK, I was just trying to get clarification on who "we" was in https://github.com/scipy/scipy/issues/18574#issuecomment-1687534004 : ) But if you can get the initial deprecation rolling, that will give us - whomever that may be - some time before more is needed.

**Comment by ilayn** [comment]
I am doing the quadpack rounds for #21201 and stumbled upon this one. Hence in case you need a gentle nudge because the translation is in place, you got it 😃

**Comment by h-vetinari** [comment]
Wow, very impressive 👏 

I'm guessing you're leaving the API (incl. return types) untouched in that PR, but we could/should iterate on this soon after?

Not sure anymore where we stand on this:
> @mdhaber: So I'd propose that we deprecate `quadrature` and `romberg`, directing users to `quad` instead.

**Comment by mdhaber** [comment]
> I'm guessing you're leaving the API (incl. return types) untouched in that PR

Yes

> Not sure anymore where we stand on this:

TBH I'm not really sure either. We could have added these methods to `_tanhsinh` and made it public by now. But apparently there are other plans. Perhaps ev-br can comment.

**Comment by ev-br** [comment]
Not that I've a detailed plan off the top of my head. Roughly, I'd think
- `quadrature` and `romberg` we can deprecate
- merging `quad`, `quad_vec` and `tanhsinh`: I'd expect that it'd be easiest to leave existing functions alone to become legacy/deprecated in some future, and cook up a new function with a sane API. Maybe it's easier now that the majority of QUADPACK workhorse functions are in C, so they can be connected to python together with tanhsinh.

**Comment by izaid** [comment]
> Not that I've a detailed plan off the top of my head. Roughly, I'd think
> 
>     * `quadrature` and `romberg` we can deprecate
> 
>     * merging `quad`, `quad_vec` and `tanhsinh`: I'd expect that it'd be easiest to leave existing functions alone to become legacy/deprecated in some future, and cook up a new function with a sane API. Maybe it's easier now that the majority of QUADPACK workhorse functions are in C, so they can be connected to python together with tanhsinh.

Yeah, basically this. We have a cubature PR incoming, and one specific intention of that is to propose an API that is flexible enough for these different use cases. If that succeeds, perhaps we can port lessons learned to the 1D quad case.

**Comment by lucascolley** [comment]
> quadrature and romberg we can deprecate

Just to make sure we're on the same page, they are already deprecated, and slated for removal next year https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.romberg.html.

**Comment by lucascolley** [comment]
let's close this now, the future of these capabilities in SciPy can be discussed in gh-20252 and beyond.

**Comment by h-vetinari** [comment]
I don't see how new API relates to this issue (except if we totally remove all traces of our old API, which would however need a long-term deprecation and clear migration story)

**Comment by mdhaber** [comment]
My thought was that if users are willing to make changes, they might as well switch to the new API completely. If they are not, they still with this old API, which we could just leave alone.

**Comment by lucascolley** [comment]
Feel free to reopen! @mdhaber's thought was that we are unlikely to do anything about this other than (perhaps eventually) making them obsolete

**Comment by h-vetinari** [comment]
Yeah, so I was surprised that the quadrature issue got referred to a new cubature API. But fine by me if there's a full replacement. IMO we should still deprecate the old API (for removal in 2.0 or so), and write up a guide how to move to the new API.

**Comment by ev-br** [comment]
`cubature` is not a `quad` replacement.

**Comment by izaid** [comment]
> `cubature` is not a `quad` replacement.

I think the hope is it, or a 1D `quadrature` variant, could be in time with a bit more optimisation?

**Comment by ev-br** [comment]
> could be in time 

+1 to that! Not ATM though.

>  with a bit more optimisation

... and a bit more algo work :-).

**Comment by h-vetinari** [comment]
I was referring to this

> they might as well switch to the new API (`quadrature`/`cubature`, IMO) completely

Not sure how much the `quadrature` API covers `quad`, but as long as we can map 99% use cases to either `quadrature` or `cubature`, that should be good enough.
